cli/src/package.rs

```diff
@@ -1,36 +1,135 @@
-use std::{env::current_dir, fs, path::PathBuf};
-
+use crate::error::CliError;
 use directories::ProjectDirs;
+use futures::future::join_all;
+use modmark_core::Resolve;
+use serde_json;
+use std::{
+    env::current_dir,
+    fs::{self, create_dir_all, File},
+    io::copy,
+    path::PathBuf,
+};
 
-pub trait Resolve {
-    fn resolve(&self, path: &str) -> Option<Vec<u8>>;
-}
+static RUNTIME: once_cell::sync::Lazy<tokio::runtime::Runtime> = once_cell::sync::Lazy::new(|| {
+    tokio::runtime::Builder::new_multi_thread()
+        .enable_all()
+        .build()
+        .unwrap()
+});
+
+pub struct PackageManager {}
 
-pub struct PackageManager;
 impl Resolve for PackageManager {
-    fn resolve(&self, path: &str) -> Option<Vec<u8>> {
-        let mut local_path = match current_dir() {
-            Ok(p) => p,
-            Err(_) => return None,
-        };
+    type Error = CliError;
+    fn resolve(&self, path: &str) -> Result<Vec<u8>, Self::Error> {
+        let out = RUNTIME.block_on(resolve_package(path))?;
+
+        return Ok(out);
+    }
+    fn resolve_all(&self, paths: Vec<&str>) -> Vec<Result<Vec<u8>, Self::Error>> {
+        let out = RUNTIME.block_on(resolve_packages(paths));
 
-        local_path.push(PathBuf::from(path));
+        return out;
+    }
+}
 
-        let mut cache_path = match ProjectDirs::from("org", "modmark", "packages") {
-            Some(path) => path.cache_dir().to_path_buf(),
-            None => return None,
+async fn resolve_packages(paths: Vec<&str>) -> Vec<Result<Vec<u8>, CliError>> {
+    let futures = paths.iter().map(|path| resolve_package(path));
+
+    return join_all(futures).await;
+}
+
+async fn resolve_package(path: &str) -> Result<Vec<u8>, CliError> {
+    let splitter = path.split_once(":");
+
+    let Some((specifier, package_path)) = splitter else {
+           return fetch_local(path);
         };
 
-        cache_path.push(PathBuf::from(path));
+    match specifier {
+        "http" => fetch_url("", path).await,
+        "https" => fetch_url("", path).await,
+        "pkgs" => fetch_registry(package_path).await,
+        other => Err(CliError::Specifier(other.to_string())),
+    }
+}
+
+async fn fetch_url(url_root: &str, package_path: &str) -> Result<Vec<u8>, CliError> {
+    let mut cache_path = match ProjectDirs::from("org", "modmark", "packages") {
+        Some(path) => path.cache_dir().to_path_buf(),
+        None => return Err(CliError::Cache),
+    };
+
+    cache_path.push(PathBuf::from(&package_path));
```

Comment:
Hmm, this means that (I have tested this) loading the url `https://github.com/modmark-org/packages/raw/main/bold.wasm` will make a cache file in `/Users/cmdjojo/Library/Caches/org.modmark.packages/https://github.com/modmark-org/packages/raw/main/bold.wasm` which I don't think is intentional; I don't think `https://` is a valid dir name and it causes confusion for MacOS. I don't think the `https://` is needed, maybe just the domain and path?

Reply:
Yeah I agree